src/item.cpp

```diff
@@ -8067,6 +8067,23 @@ std::string item::components_to_string() const
     }, enumeration_conjunction::none );
 }
 
+uint64_t item::make_component_hash() const
+{
+    // First we need to sort the IDs so that identical ingredients give identical hashes.
+    std::multiset<std::string> id_set;
+    for( const item &it : components ) {
+        id_set.insert( it.typeId() );
+    }
```

Comment:
If I understand this correctly, food with even the slightly different components will not suffer from the monotony penalty, and food with the same components but completely different sub-components will still suffer from the penalty? I think ideally there should be a metric to determine how similar two foods are, but I'm not sure how (or whether) that could be done.

Reply:
I just have another idea, though it might actually belong to a separate PR: count the fun bonus and monotony penalty from sub-components and cooking recipes separately, so making different foods from the same materials, and making the same food from different materials will both impose a penalty, while eating the same food made from the same materials gives the most penalty.

Reply:
Not a bad idea, but again, I don't see a way of doing that without storing copies of the items themselves in the `consumption_event` struct. Doing this for every NPC as well as the player would be a huge performance overhead for a relatively small feature.